[WIP] Swap thread state before forwarding to Python apply

malvika2147 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#21634 Swap thread state before forwarding to Python apply**

Summary:
Before calling into the python apply_fn from PyFunction::apply, swap to a new PyThreadState to prevent recursion error in python in the cases of reentrant backwards.

TODO : Do not swap thread state for non reentrant backward calls. 

Test Plan:
Tested with https://github.com/pytorch/pytorch/issues/6959 to ensure no Recursion Error

Commit:
Swap thread state before forwarding to Python apply

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

ezyang commented:
Note that the switching and unswitching will need to be done in RAII form. But apparently we still haven't solved the deadlock.

Commit:
Update on "[WIP] Swap thread state before forwarding to Python apply"

Swap thread state before forwarding to Python apply

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

gh-metadata: pytorch pytorch 21634 gh/mal2147/8/head

Commit:
Update on "[WIP] Swap thread state before forwarding to Python apply"

Swap thread state before forwarding to Python apply

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

gh-metadata: pytorch pytorch 21634 gh/mal2147/8/head

Commit:
Update on "[WIP] Swap thread state before forwarding to Python apply"

Swap thread state before forwarding to Python apply

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

gh-metadata: pytorch pytorch 21634 gh/mal2147/8/head

Commit:
Update on "[WIP] Swap thread state before forwarding to Python apply"

Swap thread state before forwarding to Python apply

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

gh-metadata: pytorch pytorch 21634 gh/mal2147/8/head

ezyang commented on torch/csrc/autograd/python_function.cpp:
```diff
@@ -1079,8 +1086,12 @@ bool THPFunction_initModule(PyObject *module)
   void operator()(PyFunction* p) const {
-    AutoGIL gil;
-    Py_DECREF(p->obj);
+    if (swappedThreadState) {
```
Wouldn't it be less error prone if you just had `AutoGIL` take care of deciding whether or not the GIL needs to be taken out again? Also, the Python invariants for when you're allowed to take out a GIL while you have it, and when you are not, deserve a beefy note somewhere in the code :)

ezyang replied:
Yeah, you've edited a bunch of AutoGIL sites, and I don't have very much confidence that you've necessarily gotten all of them :)

Alternately, could we perhaps make it a BUG to take out AutoGIL while you hold it, and then fix up cases where we made this mistake? (I could believe that the sites you've identified are exactly those). Something like a `TORCH_INTERNAL_ASSERT` in AutoGIL that we don't already have the GIL. This would be a reasonable patch which we could land before this one.

ezyang commented on torch/csrc/utils/swap_thread_state.h:
```diff
@@ -0,0 +1,19 @@
+#pragma once
+
+// RAII struct to change the current PyThreadState to a new PyThreadState
```
A little explanation about what a PyThreadState here would be pretty useful. Similarly, another question I like answered in docs is "When would I want to use this?"

ezyang commented on torch/csrc/utils/swap_thread_state.cpp:
```diff
@@ -0,0 +1,23 @@
+ }
+
+SwapPyThreadState::~SwapPyThreadState() {
+  // Copy exception to the old thread state
```
Interesting. Is it possible to skip doing this transference, because the autograd engine is going to catch and process the Python exception in short order? (In particular, it needs to propagate the exception back to the main thread). Though I guess it probably doesn't matter too much.

ezyang commented on torch/csrc/autograd/python_function.cpp:
```diff
@@ -133,7 +134,13 @@ auto PyFunction::apply(variable_list&& inputs) -> variable_list {
+  {
+    SwapPyThreadState swapThreadState;
+    callRes = PyObject_CallObject(apply_fn, pyInputs.get());
+  }
```
Nice :) Though I wonder why you couldn't have assigned `callRes` to the `THPObjectPtr` to start with; assignment is defined on THPObjectPtr.

ezyang commented:
This is super cool. Some high level comments:

* It took a bunch of investigation to figure out how exactly to appease the Python interpreter invariants, right? I'd love to see a nice `Note [PyThreadState invariants]` explaining the findings somewhere
* I think it may be very reasonable to say, "In PyTorch, we never take out the GIL when we have it already", and then just fix up sites where we've messed this up. This would be a good preparatory patch for this one, IMO.

Other stuff is nits. Very cool stuff, and I'm impressed all the tests pass :)

ezyang commented:
Recording for posterity: it turns out you're not allowed to swap the
PyThreadState for a given CPU thread; once you allocate a PyThreadState
for a CPU thread, you're never allowed to change it.  So this strategy
doesn't work.

Excerpts from Edward Z. Yang's message of 2019-06-14 08:12:08 -0700:
> This is super cool. Some high level comments:
> 
> * It took a bunch of investigation to figure out how exactly to appease the Python interpreter invariants, right? I'd love to see a nice `Note [PyThreadState invariants]` explaining the findings somewhere
> * I think it may be very reasonable to say, "In PyTorch, we never take out the GIL when we have it already", and then just fix up sites where we've messed this up. This would be a good preparatory patch for this one, IMO.
> 
> Other stuff is nits. Very cool stuff, and I'm impressed all the tests pass :)
>

apaszke reviewed (commented):
~~Did we benchmark how expensive it is to create new `PyState`s? We're doing this at every single call to a Python-based autograd function after all... If you're facing recursion errors, why can't you just `sys.setrecursionlimit`? That's the standard solution that will work with arbitrary Python code. I don't think it's a PyTorch problem...~~

Never mind. I hate that both merged and abandoned pull requests show up as "closed" on GitHub...

VitalyFedyunin commented on torch/csrc/utils/cuda_lazy_init.cpp:
```diff
@@ -5,24 +5,37 @@
+    // has a buggy implementation that deadlocks if an instance throws an
+    // exception.  In any case, call_once isn't necessary, because we
+    // have taken a lock.
+    static bool run_yet = false;
```
Such copy-paste is error prone in future. Imagine if someone will change one branch and forgot to change another.
Consider moving code to something like `unsafe_cuda_lazy_init()` and calling it on both sites.

VitalyFedyunin commented on torch/csrc/utils/swap_thread_state.h:
```diff
@@ -0,0 +1,19 @@
+// Check this thread local before trying to acqure the GIL using
+// PyGILStateEnsure recursively. If the ThreadState has been swapped and the GIL
+// hasn't been released, do not try to acquire it again from the same thread
+extern thread_local bool swappedThreadState;
```
Consider using namespaces
